dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -824,36 +848,37 @@ bool Aggregator::executeOnBlock(
         && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by)
         && worth_convert_to_two_level)
     {
-        writeToTemporaryFile(result, file_provider);
+        /// todo: the memory usage is calculated by memory_tracker, it is not accurate since memory tracker
+        ///  will tracker all the memory usage for the task/query, need to record and maintain the memory usage
+        ///  in Aggregator directly.
+        spill(result);
     }
 
     return true;
 }
 
 
-void Aggregator::writeToTemporaryFile(AggregatedDataVariants & data_variants, const FileProviderPtr & file_provider)
+void Aggregator::finishSpill()
 {
-    Stopwatch watch;
-    size_t rows = data_variants.size();
-
-    auto file = std::make_unique<Poco::TemporaryFile>(params.tmp_path);
-    const std::string & path = file->path();
-    WriteBufferFromFileProvider file_buf(file_provider, path, EncryptionPath(path, ""));
-    CompressedWriteBuffer compressed_buf(file_buf);
-    NativeBlockOutputStream block_out(compressed_buf, ClickHouseRevision::get(), getHeader(false));
-
-    LOG_DEBUG(log, "Writing part of aggregation data into temporary file {}.", path);
+    assert(spiller != nullptr);
+    spiller->finishSpill();
+}
 
-    /// Flush only two-level data.
+BlockInputStreams Aggregator::restoreSpilledData()
+{
+    assert(spiller != nullptr);
+    return spiller->restoreBlocks(0);
+}
 
-#define M(NAME)                                                                   \
-    case AggregationMethodType(NAME):                                             \
-    {                                                                             \
-        writeToTemporaryFileImpl(                                                 \
-            data_variants,                                                        \
-            *ToAggregationMethodPtr(NAME, data_variants.aggregation_method_impl), \
-            block_out);                                                           \
-        break;                                                                    \
+void Aggregator::spill(AggregatedDataVariants & data_variants)
+{
+    /// Flush only two-level data and possibly overflow data.
```

Comment:
So currently `extern-agg-merge` is enabled by default?

Reply:
No, the default value of `max_bytes_before_external_group_by` is 0, which means no external agg.